src/AppInstallerRepositoryCore/Rest/RestSource.cpp

```diff
@@ -0,0 +1,272 @@
+
+        // The source reference used by package objects.
+        struct SourceReference
+        {
+            SourceReference(const std::shared_ptr<const RestSource>& source) :
+                m_source(source) {}
+
+        protected:
+            std::shared_ptr<const RestSource> GetReferenceSource() const
+            {
+                std::shared_ptr<const RestSource> source = m_source.lock();
+                THROW_HR_IF(E_NOT_VALID_STATE, !source);
+                return source;
+            }
+
+        private:
+            std::weak_ptr<const RestSource> m_source;
+        };
+
+        // The IPackageVersion impl for RestSource.
+        struct PackageVersion : public SourceReference, public IPackageVersion
+        {
+            PackageVersion(
+                const std::shared_ptr<const RestSource>& source, IRestClient::PackageInfo packageInfo, VersionAndChannel versionInfo)
+                : SourceReference(source), m_packageInfo(packageInfo), m_versionInfo(versionInfo) {}
+
+            // Inherited via IPackageVersion
+            Utility::LocIndString GetProperty(PackageVersionProperty property) const override
+            {
+                switch (property)
+                {
+                case PackageVersionProperty::SourceIdentifier:
+                    return Utility::LocIndString{ GetReferenceSource()->GetIdentifier() };
+                case PackageVersionProperty::SourceName:
+                    return Utility::LocIndString{ GetReferenceSource()->GetDetails().Name };
+                case PackageVersionProperty::Id:
+                    return Utility::LocIndString{ m_packageInfo.packageIdentifier };
+                case PackageVersionProperty::Name:
+                    return Utility::LocIndString{ m_packageInfo.packageName };
+                case PackageVersionProperty::Version:
+                    return Utility::LocIndString{ m_versionInfo.GetVersion().ToString() };
+                case PackageVersionProperty::Channel:
+                    return Utility::LocIndString{ m_versionInfo.GetChannel().ToString() };
+                default:
+                    return Utility::LocIndString{};
+                }
+            }
+
+            // TODO
+            std::vector<Utility::LocIndString> GetMultiProperty(PackageVersionMultiProperty property) const override
+            {
+                UNREFERENCED_PARAMETER(property);
+                std::vector<Utility::LocIndString> result;
+                return result;
+            }
+
+            // TODO
+            Manifest::Manifest GetManifest() const override
+            {
+                AICLI_LOG(Repo, Verbose, << "Downloading manifest");
+                Manifest::Manifest manifest = GetReferenceSource()->GetRestClient().GetManifestByVersion(
+                    m_packageInfo.packageIdentifier, m_versionInfo.GetVersion().ToString(), m_versionInfo.GetChannel().ToString()).value();

```

Comment:
>.value() [](start = 130, length = 8)

std::optional check it's set before calling Value()

Reply:
you are right, I will just receive it as std::optional. I will implementing and using this in my next PR.

---
In reply to: [594733374](https://github.com/microsoft/winget-cli/pull/765#discussion_r594733374) [](ancestors = 594733374)